### ISSUE TITLE
Decoding with different batch sizes yields different results.

### ISSUE THREAD

**Issue by helson73** [OpenNMT/OpenNMT-py#502]
Translation batch size 50:
BLEU = 25.76, **61.1**/32.5/19.2/**11.7** (BP=**0.998**, ratio=**0.998**, hyp_len=**44447**, ref_len=44557)

Translation batch size 80:
BLEU = 25.76, **61.0**/32.5/19.2/**11.6** (BP=**0.999**, ratio=**0.999**, hyp_len=**44491**, ref_len=44557)

How could this happen?
My option is:
python translate.py -gpu 0 -verbose -replace_unk -alpha 0.6 -beam_size 5 -beta 0.2

**Comment by srush** [comment]
Can you do file diff? This is such a small difference that it's going to be hard to track down. 

Batch beam search is one of the trickier parts of the code. I think it is pretty solid, but there might be a corner case

**Comment by helson73** [comment]
I can't find the original file, so I re-run on a different epoch save.

Batch 50:
BLEU = 26.13, 61.2/32.8/19.5/11.9 (BP=1.000, ratio=1.007, hyp_len=**44880**, ref_len=44557)
Batch 80:
BLEU = 26.13, 61.2/32.8/19.5/11.9 (BP=1.000, ratio=1.007, hyp_len=**44870**, ref_len=44557)

Note that bleu was tested on '@@' restored text, and diff is done on original files (bpe).

`diff results/pred_ep28_bs50.en results/pred_ep28_bs80.en > diff.log`

> 99c99
> < the worldwide research competition was started , especially in Japan , because the experiment of high temperature superconductivity of yt@@ tri@@ um and copper oxide began in 1987 , and the promotion policy was taken by the country so that it does not delay in Japan .
> ---
> > the worldwide research competition was started , especially in Japan , because the experiment of high temperature superconductivity of yt@@ tri@@ um and copper oxide began in 1987 , and the promotion policy was entr@@ u@@ sted by the country so that it does not delay in Japan .
> 276c276
> < the advantage is that there is no seal part , and the safety is high , and there is no ex@@ cur@@ sion in the tube , and the sensor can exchange the flow of the process even in the case of the trouble , etc ..
> ---
> > the advantage is that there is no seal part , and the safety is high , and there is no ex@@ cur@@ sion in the tube , and the sensor can exchange the flow of the process even in the trouble , because there is outside the pipe .
> 458c458
> < ta@@ urine intake of a T 98 cell had a high affinity system in Na + dependence , and decreased by administration of az@@ obenz@@ ene 12 , 13@@ -1@@ 00@@ 00 ( I , 100 ) and iso@@ prote@@ reno@@ l ( 1@@ μM ) .
> ---
> > ta@@ urine intake of a T 98 cell had a high affinity system in Na + dependence , and decreased by administration of az@@ obenz@@ ene 12 , 13@@ -1@@ 00@@ 00 ( I , 100@@ g ) and iso@@ prote@@ reno@@ l ( 1@@ μM ) .
> 644c644
> < it is considered that the mucosa damaged by the parasi@@ tism of C@@ amp@@ ylo@@ bac@@ tor , and the environment suitable for growth and breeding of C@@ lo@@ stri@@ dium is produced , and the ship@@ ment cattle died , and the ship@@ ment cattle may cause the necrotic enteritis .
> ---
> > it is considered that the mucosa damaged by the parasi@@ tism of the C@@ amp@@ ylo@@ bac@@ tor , and the environment suitable for growth and breeding of C@@ lo@@ stri@@ dium is produced , and the ship@@ ment cattle is caused by the necro@@ tizing enteritis .
> 854c854
> < though the uterine leiomy@@ oma was the most suspected by various examinations such as pelvic part MRI , total hy@@ ste@@ rectomy by g@@ yn@@ a@@ ec@@ olog@@ ia and bilateral ad@@ n@@ ex@@ ectomy by g@@ yn@@ a@@ ec@@ olog@@ ia were carried out by the preoperative diagnosis that malignancy could not den@@ ied .
> ---
> > though the uterine leiomy@@ oma was the most suspected by various examinations such as pelvic part MRI , total hy@@ ste@@ rectomy by g@@ yn@@ a@@ ec@@ olog@@ ia and bilateral ad@@ n@@ ex@@ ectomy by g@@ yn@@ a@@ ec@@ olog@@ ia were carried out by the preoperative diagnosis that malignancy could not be den@@ ied .
> 867c867
> < it has been reported that West N@@ ile virus ( N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ V ) N@@ N@@ N@@ N@@ N@@ P@@ 4@@ B protein becomes antagonist of interferon together with other viral proteins , and localization in cell nucleus in the later stage of infection .
> ---
> > it has been reported that West N@@ ile virus ( N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ N@@ V ) N@@ N@@ N@@ N@@ N@@ P@@ 4@@ B protein becomes antagonist of interferon together with other viral proteins , and that it is localized in the cell nucleus in the later stage of the infection .
> 1270c1270
> < on the trend of the engine@@ er , the following are described : demand and supply quantity , change of demand structure , composition ratio by man and woman , employment trend of gradu@@ ate by gradu@@ ate university , and trend of the w@@ ill@@ age technology , etc ..
> ---
> > on the trend of the engine@@ er , the following are described : demand and supply quantity , change of demand structure , composition ratio by man and woman , employment trend of gradu@@ ate by gradu@@ ate university , and trend of the w@@ ill@@ age technology .
> 1273c1273
> < this report focuses on hand@@ writing vo@@ te and public disc@@ losure card as information service medium in public occupational stable office , and it is easy to carry out the experiment using recognition psych@@ ology or recognition engineering technique " which is easy to carry out " experiment " .
> ---
> > this report focuses on hand@@ writing vo@@ te and public disc@@ losure card as information service medium in public occupational stable office , and it is easy to carry out the experiment using recognition psych@@ ology or recognition engineering technique " which is easy to carry out .
> 1280c1280
> < in the fuel system@@ atization system of the sew@@ er@@ age industry in Japan , the conventional drying system and carbon@@ ization system are optimized for fuel production , and the fuel consumption improvement in operation is aimed at the same time as the high calori@@ e of generated fuel .
> ---
> > in the fuel system@@ atization system of the sew@@ er@@ age industry in Japan , the conventional drying system and carbon@@ ization system are optimized for fuel production , and simultaneously , fuel consumption improvement in operation is aimed at the same time .
> 1386c1386
> < in order to improve the starting characteristics of a water motor pump for pumping , a variable reactor starting device capable of changing the power supply voltage to three stages was developed , and a comparison experiment was carried out with direct start .
> ---
> > in order to improve the starting characteristics of a water motor pump for pumping , a variable reactor starting device capable of changing the power supply voltage to three stages was developed , and a comparison experiment was carried out with direct injection start .
> 1516c1516
> < the object grinding machine is called a cr@@ ushing machine , and the particle is accelerated by the high-speed jet , and cr@@ ushing the particle by the high-speed jet , and cr@@ ushing it by the wall surface of the particle and particle .
> ---
> > the object grinding machine is called a cr@@ ushing machine , and the particle is accelerated by the high-speed jet , and cr@@ ushing the particle by the high-speed jet , and cr@@ ushing it by the collision of the particle and wall surface of the particle is carried out .
> 1665c1665
> < although the air@@ frame was immediately modified to the run@@ way direction , the left main leg cont@@ acted to the left side run@@ way light without rising , and in addition , it was extended from the run@@ way end to about 500@@ 00@@ 00 .
> ---
> > although the air@@ frame was immediately modified to the run@@ way direction , the left main leg cont@@ acted to the left side run@@ way light without rising , and in addition , it was extended from the run@@ way end to about 500@@ 00@@ 00 in the extension direction .
> 1696c1696
> < though there is the constraint in the application by the delay time of the processing system in the active noise control until now , the prediction of the noise becomes possible , if the property of the noise source is utilized , etc ..
> ---
> > though there is the constraint in the application by the delay time of the processing system in the active noise control until now , the prediction of the noise becomes possible , if the property of the noise source is utilized , etc . , it becomes possible .
> 1805c1805
> < the light visual field image and dark field image observation were carried out at a magn@@ ification of several tens of magnitude , and the equal thickness interference fring@@ e corresponding to the change of thickness was observed .
> ---
> > the light visual field image and dark field image observation were carried out at a magn@@ ification of several tens of magnitude , and the equal thickness interference fring@@ e corresponding to the change in thickness of the sample surface was observed .

**Comment by helson73** [comment]
@srush 
I am wondering shouldn't layernorm turn off when it's on validation or test ?
```python
class LayerNorm(nn.Module):
    ''' Layer normalization module '''

    def __init__(self, d_hid, eps=1e-3):
        super(LayerNorm, self).__init__()

        self.eps = eps
        self.a_2 = nn.Parameter(torch.ones(d_hid), requires_grad=True)
        self.b_2 = nn.Parameter(torch.zeros(d_hid), requires_grad=True)

    def forward(self, z):
        if z.size(1) == 1 or **not self.train**:
            return z
        mu = torch.mean(z, dim=1)
        sigma = torch.std(z, dim=1)
        # HACK. PyTorch is changing behavior
        if mu.dim() == 1:
            mu = mu.unsqueeze(1)
            sigma = sigma.unsqueeze(1)
        ln_out = (z - mu.expand_as(z)) / (sigma.expand_as(z) + self.eps)
        ln_out = ln_out.mul(self.a_2.expand_as(ln_out)) \
            + self.b_2.expand_as(ln_out)
        return ln_out
```

**Comment by srush** [comment]
I think it does turn off `not self.train`

Are you seeing otherwise?

**Comment by srush** [comment]
No, I think it is supposed to remain on: 

"""Like batch normalization,
we also give each neuron its own adaptive bias and gain which are applied after
the normalization but before the non-linearity. Unlike batch normalization, layer
normalization performs exactly the same computation at training and test times."""

**Comment by helson73** [comment]
Oh, my mistake, I thought the second dimension is batch.

**Comment by srush** [comment]
oh, it might be wrong. this is a smart place to check. Can you look into what size it is getting?

**Comment by helson73** [comment]
It is <batch_size, len, dim>.

**Comment by srush** [comment]
Oh, so you think it is incorrect? Either way this module needs much clear documentation and tests.

**Comment by srush** [comment]
maybe we should switch this code to use a different layernorm impl. There are a couple here https://github.com/pytorch/pytorch/issues/1959

**Comment by srush** [comment]
Okay, so because we use Bottle the dimensions in layer norm are actually: 

<batch_size x beam_size x len, dim>

So in theory this will change based on how many batches are used. However, I can't tell if that is incorrect or not. I need to reread the layer norm paper.
